Libraries/PeriphDrivers/Source/SPI/spi_reva2.c

```diff
@@ -1417,12 +1417,13 @@ int MXC_SPI_RevA2_ControllerTransaction(mxc_spi_reva_regs_t *spi, uint8_t *tx_bu
     MXC_SPI_RevA2_transactionSetup(spi, tx_buffer, tx_length_frames, rx_buffer, rx_length_frames,
                                    false);
 
-    // Start the SPI transaction.
-    spi->ctrl0 |= MXC_F_SPI_REVA_CTRL0_START;
-
     // Handle Target Select Pin (Only applicable in HW_AUTO TS control scheme).
     MXC_SPI_RevA2_handleTSControl(spi, deassert, hw_ts_index);
 
+
+    // Start the SPI transaction.
+    spi->ctrl0 |= MXC_F_SPI_REVA_CTRL0_START;
+
```

Comment:
Moving this after the TS control logic could cause a regression related to glitches on the TS line.

As soon as the `CTRL0->START` bit is set, the hardware will reinitialize the TS pin based on the value of `CTRL0->SS_CTRL`.  It will cause a small glitch if:
- The TS line is asserted
- We want to deassert the line as part of this transaction

I saw this during fastspi.c development for SRAM drivers [here](https://github.com/Analog-Devices-MSDK/msdk/blob/bcdd4695d596e5b714d5e3aba3b2045cda58649e/Libraries/MiscDrivers/SRAM/fastspi.c#L296)

I would suggest we revert back to setting the START bit first, then handling the TS control logic.  Did you find an issue where the new order is necessary?

Reply:
Thanks @ozan956, maybe we need to re-examine the logic for the [handleTSControl](https://github.com/Analog-Devices-MSDK/msdk/blob/bcdd4695d596e5b714d5e3aba3b2045cda58649e/Libraries/PeriphDrivers/Source/SPI/spi_reva2.c#L1331) implementation.

Do you happen to have a capture showing the failed first transaction?

The scenario for the regression test is when devices need half-duplex SPI transactions.  An example is below.

![TS glitch](https://github.com/Analog-Devices-MSDK/msdk/assets/38844790/8622c6ed-5267-47a3-acf4-0f0117b5ca3a)
([N01S830HA datasheet](https://www.onsemi.com/pdf/datasheet/n01s830ha-d.pdf))

The only way to achieve this with our SPI hardware is to run two consecutive transactions, while keeping TS asserted in between.  If the START bit is set _after_ the deassertion logic is handled, then we will see the red glitch in the image above, which can cause communication issues with slave devices.

Reply:
Hello Jake, I will provide the screenshots of the logic tomorrow.

Reply:
hi @Jake-Carter, have you had a change to check this ?

Reply:
Thanks @MertCanBoz1 @sdivarci, sorry for the delay.  In the future tag me in the comment so I get the notification, otherwise I get so many I might not see it :)

You're right - it looks like the `handleTSControl` function introduces a ~20uS delay...  not good. 

If I remember correctly, I believe the deassert logic is the critical thing to avoid the glitch.  Can re-test with this approach?
1. Keep your change to call `spi->ctrl0 |= MXC_F_SPI_REVA_CTRL0_START;` _after_ `MXC_SPI_RevA2_handleTSControl`.
2. Move the deassert logic _out_ of `MXC_SPI_RevA2_handleTSControl`.
3. Place it _after_ the start of the transaction.

```diff
diff --git a/Libraries/PeriphDrivers/Source/SPI/spi_reva2.c b/Libraries/PeriphDrivers/Source/SPI/spi_reva2.c
index 9af1be7768..e872905c7e 100644
--- a/Libraries/PeriphDrivers/Source/SPI/spi_reva2.c
+++ b/Libraries/PeriphDrivers/Source/SPI/spi_reva2.c
@@ -1348,12 +1348,6 @@ static void MXC_SPI_RevA2_handleTSControl(mxc_spi_reva_regs_t *spi, uint8_t deas
         // Limitation: This implemention only support transactions with one target at a time.
         MXC_SETFIELD(spi->ctrl0, MXC_F_SPI_REVA_CTRL0_SS_ACTIVE,
                      ((1 << hw_ts_index) << MXC_F_SPI_REVA_CTRL0_SS_ACTIVE_POS));
-
-        if (deassert) {
-            spi->ctrl0 &= ~MXC_F_SPI_REVA_CTRL0_SS_CTRL;
-        } else {
-            spi->ctrl0 |= MXC_F_SPI_REVA_CTRL0_SS_CTRL;
-        }
     }
 
     // Add support for SW_DRV TS Control here in the future.
@@ -1387,11 +1381,17 @@ int MXC_SPI_RevA2_ControllerTransaction(mxc_spi_reva_regs_t *spi, uint8_t *tx_bu
     MXC_SPI_RevA2_transactionSetup(spi, tx_buffer, tx_length_frames, rx_buffer, rx_length_frames,
                                    false);
 
+    // Handle Target Select Pin (Only applicable in HW_AUTO TS control scheme).
+    MXC_SPI_RevA2_handleTSControl(spi, deassert, hw_ts_index);
+
     // Start the SPI transaction.
     spi->ctrl0 |= MXC_F_SPI_REVA_CTRL0_START;
 
-    // Handle Target Select Pin (Only applicable in HW_AUTO TS control scheme).
-    MXC_SPI_RevA2_handleTSControl(spi, deassert, hw_ts_index);
+    if (deassert) {
+        spi->ctrl0 &= ~MXC_F_SPI_REVA_CTRL0_SS_CTRL;
+    } else {
+        spi->ctrl0 |= MXC_F_SPI_REVA_CTRL0_SS_CTRL;
+    }
 
     // Complete transaction once it started.
     while (STATES[spi_num].transaction_done == false) {
```

Reply:
Hi @Jake-Carter. I've tried the approach, and it works. 

Here is the result: 
![image](https://github.com/Analog-Devices-MSDK/msdk/assets/49796358/c69c2707-d8db-4549-a3da-632080c14749)

After your approval, I can also commit the changes in the "spi_reva2.c".

Reply:
Thanks @ozan956, yes please do.  Also be sure to move the comment about the deassert logic:

```C
    // Handle target-select (L. SS) deassertion. This must be done
    //   AFTER launching the transaction to avoid a glitch on the TS line if:
    //     - The TS line is asserted
    //     - We want to deassert the line as part of this transaction
    //
    // As soon as the SPI hardware receives CTRL0->START it seems to reinitialize the Target Select (TS) pin based
    //   on the value of CTRL->SS_CTRL, which causes the glitch.
    if (deassert) ...

Reply:
Hi @Jake-Carter, I made the changes and re-requested your review.